addons/sale_timesheet/models/sale_order.py

```diff
@@ -72,6 +73,25 @@ def _compute_field_value(self, field):
                 # We want to display only one time the warning for each SOL
                 upsellable_lines.write({'has_displayed_warning_upsell': True})
 
+    def _compute_show_hours_recorded_button(self):
+        orders_to_check = self.filtered(lambda l: l.state not in ['draft', 'sent'])
+        list_order = self.env['sale.order.line']._read_group([('order_id', 'in', orders_to_check.ids)], ['product_id:array_agg(product_id)'], ['order_id'], lazy=False)
+        dict_order = {}
+        sol_product_ids = set()
+        for order in list_order:
+            product_set = set(order['product_id'])
+            dict_order[order['order_id'][0]] = product_set
+            sol_product_ids.update(product_set)
+
+        products_service_prepaid_or_timesheet = self.env['product.product'].search([('id', 'in', list(sol_product_ids)), ('type', '=', 'service'),
+                                                                                    '|', ('service_type', 'not in', ['milestones', 'manual']), ('invoice_policy', '!=', 'delivery')])
+        product_service_prepaid_or_timesheet_ids = set(products_service_prepaid_or_timesheet.ids)
+        for order in self:
+            if order in orders_to_check and not product_service_prepaid_or_timesheet_ids.isdisjoint(dict_order[order.id]):
+                order.show_hours_recorded_button = True
+            else:
+                order.show_hours_recorded_button = False
```

Comment:
You could directly use the condition like this:
```suggestion
           order.show_hours_recorded_button = order in orders_to_check and not product_service_prepaid_or_timesheet_ids.isdisjoint(dict_order[order.id]):
```